kernel/src/platform/system_timer.rs

```diff
@@ -0,0 +1,164 @@
+//! Interface for use by the Kernel to configure timers which can preempt userspace
+//! processes.
+
+use crate::hil::time::{self, Frequency};
+
+/// Interface for the system tick timer.
+///
+/// A system tick timer provides a countdown timer to enforce process scheduling
+/// quantums.  Implementations should have consistent timing while the CPU is
+/// active, but need not operate during sleep.
+///
+/// On most chips, this will be implemented by the core (e.g. the ARM core systick), but
+/// some chips lack this optional peripheral, in which case it might be
+/// implemented by another timer or alarm controller, or require virtualization
+/// on top of a single hardware timer.
+pub trait SystemTimer {
+    /// Sets the timer as close as possible to the given interval in
+    /// microseconds, and starts counting down. Interrupts are not enabled.
+    ///
+    /// Callers can assume at least a 24-bit wide clock. Specific timing is
+    /// dependent on the driving clock. For ARM boards with a dedicated
+    /// SysTick peripheral, increments of 10ms are most
+    /// accurate thanks to additional hardware support for this value, but
+    /// values up to 400ms are valid.
+    fn start_timer(&self, us: u32);
+
+    /// Returns if there is at least `us` microseconds left
+    fn at_least_us_remaining(&self, us: u32) -> bool;
+
+    /// Returns true if the timer has expired since the last time this or set_timer()
+    /// was called. If called a second time without an intermittent call to set_timer(),
+    /// the return value is unspecified (implementations can return whatever they like)
+    fn expired(&self) -> bool;
+
+    /// Resets the timer
+    ///
+    /// Resets the timer to 0 and disables it
+    fn reset(&self);
+
+    /// Disarm the underlying timer. This does not stop the timer,
+    /// but may disable the underlying interrupt if one has been set,
+    /// preventing overhead from the timer firing after an executing application
+    /// has returned to the kernel.
+    fn disarm(&self);
+
+    /// Arm the underlying timer. This does not start the timer,
+    /// it just guarantees that an interrupt will be generated
+    /// if an already started timer expires, which is useful for
+    /// preempting userspace applications.
+    fn arm(&self);
```

Comment:
Why not `enable_interrupt()`?

Reply:
@phil-levis didn't like the name `enable_interrupt()` because it might imply that all interrupts for the underlying hardware timer are disabled, which is not the case if the trait is on top of a virtualized timer.

Reply:
Ok overall I think the symmetry in arm/disarm probably makes for the most clear interface, since disarm doesn't _have_ to actually disable anything. I think just capturing the importance of this interrupt in the documentation better to explain this subtlety will avoid any questions along this line in the future. I can add a commit.